libbpf-tools/stackscan.bpf.h

```diff
@@ -0,0 +1,268 @@
+ */
+
+/*
+ * This macro can be defined to change the default before including this helper.
+ * This value is used to set the resource size for stack scan.$
+ *
+ * #define STACK_SIZE		128  Stack storage size to store per user stack
+ * #define MAX_RANGE_NR		15   Maximum number of text ranges to save
+ * #define MAX_ADDR_NR		32   Maximum number of addresses to store per user stack
+ */
+#if !defined(STACK_SIZE)
+#define STACK_SIZE		128
+#endif
+
+#if !defined(MAX_RANGE_NR)
+#define MAX_RANGE_NR		16
+#endif
+
+#if !defined(MAX_ADDR_NR)
+#define MAX_ADDR_NR		32
+#endif
+
+#define min(a, b)	(a < b ? a : b)
+
+#ifdef DEBUG
+#ifdef bpf_dbg_printk
+#error bpf_dbg_printk cannot be redefinded.
+#endif
+
+#define bpf_dbg_printk(fmt, ...) bpf_printk(fmt, ##__VA_ARGS__)
+#else
+#define bpf_dbg_printk(fmt, ...) ;
+#endif
+
+/*
+ * Limit maximum number of iterations with bpf resource limit.
+ */
+/* The maximum number of iterations of the scan that
+ * does not exceed the insn limit of the bpf program.
+ */
+#define MAX_SCAN_NR		10
+/* The maximum number of iterations of the read VMA that
+ * does not exceed the insn limit of the bpf program.
+ */
+#define MAX_VMA_NR		10
+
+/*
+ * vma
+ */
+#define VM_EXEC         0x00000004  /* include/linux/mm.h */
+
+
+#if defined(__TARGET_ARCH_arm64)
+#define IS_BL(inst)     (((inst) & 0xfc000000) == 0x94000000)
+#define IS_BLR(inst)    (((inst) & 0xfffffc1f) == 0xd63f0000)
+#define IS_BLRAA(inst)  (((inst) & 0xfffffc00) == 0xd73f0800)
+#define IS_BLRAAZ(inst) (((inst) & 0xfffffc1f) == 0xd63f081f)
+#define IS_BLRAB(inst)  (((inst) & 0xfffffc00) == 0xd73f0c00)
+#define IS_BLRABZ(inst) (((inst) & 0xfffffc1f) == 0xd63f0c1f)
+#define IS_BL_GROUP(code)	(IS_BL(code) ||     \
+				 IS_BLR(code) ||    \
+				 IS_BLRAA(code) ||  \
+				 IS_BLRAAZ(code) || \
+				 IS_BLRAB(code) ||  \
+				 IS_BLRABZ(code))
+#endif
+
+/*
+ * bpf maps for stack scanning
+ */
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(key_size, sizeof(u32));
+	__type(value, __u64[MAX_ADDR_NR]);
+	__uint(max_entries, MAX_ENTRIES);
+} scan_map SEC(".maps");
+
+/* Temporary buffer to compensate for stack limit */
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__uint(max_entries, 1);
+	__type(key, u32);
+	__type(value, STACK_SIZE);
+} stack_storage_map SEC(".maps");
+
+/* Temporary buffer to compensate for stack limit */
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__uint(max_entries, 1);
+	__type(key, u32);
+	__type(value, __u64[MAX_ADDR_NR]);
+} addrs_storage_map SEC(".maps");
+
+static __always_inline
+bool is_code_addr(u64 addr, const unsigned long ranges[][2], int nr_range)
+{
+	int i;
+
+	for (i = 0; ranges[i][0] && i < nr_range; i++) {
+		if (ranges[i][0] <= addr && addr <= ranges[i][1])
+			return true;
+	}
+
+	return false;
+}
+
+static void get_code_ranges(struct mm_struct *mm, unsigned long ranges[][2], int nr)
+{
+	int i = 0;
+	int j;
+
+	if (mm) {
+		struct vm_area_struct *vma = BPF_CORE_READ(mm, mmap);
```

Comment:
Since kernel 6.1, the mm->mmap is gone. See
  https://lore.kernel.org/all/20220906194824.2110408-9-Liam.Howlett@oracle.com/T/#me3cd1d06e86403c38419782d2c01744f2f1a0f33

The link list becomes a maple tree. The whole kernel patch series is here:
  https://lore.kernel.org/all/20220906194824.2110408-1-Liam.Howlett@oracle.com/

You will need an alternative to do this for kernel >= 6.1. I suspect some kernel change (e.g. extending bpf_find_vma helper) is needed to traverse the mm maple tree.

cc @davemarchevsky